### DISCUSSION TITLE
Suggestion: Missing Translation as Info

### DISCUSSION THREAD

**Discussion by ChristianHovenbitzer** [StefanMaron/BusinessCentral.LinterCop#804] [New Rule]
Not sure if it is possible to get translation via the language and check if any are missing, but I think this could be very beneficial.

Maybe you can share some insight? @Arthurvdv

**Comment by rvanbekkum** [comment]
Yes, it's possible. I think the only annoying thing is that I don't think it's possible for the diagnostics to point to the translation file (line) that has an issue. The ReportDiagnostic method in the Dynamics.CodeAnalysis dlls don't support the same options to raise diagnostics on (specific lines of) other file types like Roslyn allows to.
Alternatively a diagnostic could even be raised on the Caption/Label/... symbol/syntax node relevant to the issue. It is possible to determine the relevant symbol based on the "XLIFF Generator" note in the .xlf files.

**Comment by ChristianHovenbitzer** [comment]
I have started doing some development and created a draft PR just to make sure people know about this before starting to implement this themselves.

As a proof of concept this works already.
![grafik](https://github.com/user-attachments/assets/0d4ba83c-f34b-4451-bed3-8a95a0f2b5fc)

https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/884

**Reply by Arthurvdv** [reply]
This look promising, looking forward to use this rule in our pipelines.

If I can assist on anything, don't hesitate to reach out 🙋‍♂️

**Reply by ChristianHovenbitzer** [reply]
I ran into some issues already, specifically the locked property on syntax nodes like page fields. I haven't been able to figure out how to get this property
https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/5195353c5cbfe644cc8a6f3b2554a3ccaa2e0a95/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L159C8-L159C9

And accessing captions &  tooltips on request pages.
https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/5195353c5cbfe644cc8a6f3b2554a3ccaa2e0a95/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L217C13-L217C124

Any hints would be much appreciated

**Reply by rvanbekkum** [reply]
I think for the `Locked` (and `MaxLength`) subproperties you could check the `ExtractSubproperties` method.
See: https://github.com/StefanMaron/BusinessCentral.LinterCop/blob/2c1bd57e06ee75cde5213a6dcc5e0f7b22643548/BusinessCentral.LinterCop/Design/Rule0014PermissionSetCaptionLength.cs#L35

Might be good to move this to a helper codeunit, so that logic is reused.

**Reply by Arthurvdv** [reply]
![image](https://github.com/user-attachments/assets/024c4a30-26c8-488d-afd1-7a24d64d65e7)

I've added `SymbolKind.RequestPage:` to https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/5195353c5cbfe644cc8a6f3b2554a3ccaa2e0a95/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L108, where then the diagnostic also was raised on the caption on the request page. Or did you meant another example for Request Pages?

**Reply by ChristianHovenbitzer** [reply]
what about fields on the request page?

**Reply by Arthurvdv** [reply]
![image](https://github.com/user-attachments/assets/e7063299-6160-4f49-b71b-e0e664434047)

In the `GetFlattenedControls` method, a requestpage objects flows though [`IPageBaseTypeSymbol page => page.FlattenedControls`](https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/5195353c5cbfe644cc8a6f3b2554a3ccaa2e0a95/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L214) it seems, where I'm guessing you don't need a [IRequestPageBaseTypeSymbol](https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/5195353c5cbfe644cc8a6f3b2554a3ccaa2e0a95/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L217C16-L217C43)?

The [`e.RelatedFieldSymbol != null)`](https://github.com/ChristianHovenbitzer/BusinessCentral.LinterCop/blob/LC0075-LabelsShouldBeTranslated/BusinessCentral.LinterCop/Design/Rule0088LabelsShouldBeTranslated.cs#L120) causes this field to be excluded  in the `pageFields` variable though. it is not clear to me why this filter is needed? 🤔

**Reply by ChristianHovenbitzer** [reply]
I see, good catch. I think I copied that from [LC0064](https://github.com/StefanMaron/BusinessCentral.LinterCop/blob/2c1bd57e06ee75cde5213a6dcc5e0f7b22643548/BusinessCentral.LinterCop/Design/Rule0064UseTableFieldToolTip.cs#L32C41-L32C76).

But that was some helpful input, I'll look into that tomorrow

**Comment by pri-kise** [comment]
There is a new warning in the Latest AL Language Preview when the xliff has changed.
Have you checked this warning already?

**Reply by ChristianHovenbitzer** [reply]
That could resolve an issue I currently have no solution for. There needs to be some sort of caching for the xliff files, reading and parsing the entire file for every caption is not an option. I'll look into this rule, and how they check if the file has changed.

Thanks for the hint.

**Comment by ChristianHovenbitzer** [comment]
### **Table**
- **Caption**
- **Field**
  - **Caption**
  - **ToolTip**


### **Page**
- **Caption**
- **Group:**
  - **Caption**
- **Field**
  - **Caption**
  - **ToolTip**
- **Action:**
    - **Caption**
    - **ToolTip**


### **Query**
- **Caption**
- **Column**
  - **Caption**


### **Profile**
- **Caption**


### **PermissionSet**
- **Caption**


### **TableExtension**
- **Field:**
  - **Caption**
  - **ToolTip**


### **PageExtension:**
- **Layout Group**
  - **Caption**
  - **Field:**
    - **Caption**
    - **ToolTip**

### **ReportExtension**
- **RequestPage Group:**
  - **Caption**
- **Field:**
  - **Caption**
  - **ToolTip**
- **Labels:**
  - **Test**

### Enum
- Value
  - Caption

### Label Variables

**Comment by ChristianHovenbitzer** [comment]
Looks like extension objects behave different from base objects. In xliff files they dont get a translation like this:
`<trans-unit id="PageExtension 438909893 - Action 1135900470 - Property 1295455071" ..` but rather this
`<trans-unit id="Page 438909893 - Action 1135900470 - Property 1295455071" ..`
the difference being that the first id still points to the original object + name. 
@Arthurvdv  do you have an idea on how to get the base object type + name for this? I have tried a few thinks with getting the proper syntax node, but no luck so far

The LanguageFileUtilities also has a method to create this kind of id but it gets a PropertyTypeInfo, but I think it would also be feasible to just replace the first id afterwards.

**Reply by Arthurvdv** [reply]
![image](https://github.com/user-attachments/assets/4a23ea33-411e-4cc0-9802-87e41787e17c)

I took a quick look and from the Symbol, the most convenient to get the original Page the PageExtension is extending.

```C#
var pageObj = ctx.Symbol.GetContainingApplicationObjectTypeSymbol();
if (pageObj is IPageExtensionTypeSymbol pageExtObj)
{
    var originalPage = pageExtObj.Target;
}
```

@ChristianHovenbitzer , hopefully, this helps, but let me know if you need more information.

**Comment by ChristianHovenbitzer** [comment]
I think the rule is mostly stable now. There are a few things left like teaching tips but I don't see this as a major issue (lets be honest, who uses teching tips anyways 😅).
There is one bug that I don't know how to resolve however. For some background: for extension objects the rule currently compares the app id of the target object against the app id of the current object to see if the name of the **target** or **source** is relevant for the translation id. 
This logic breaks when there are multiple extension objects on the same target.


In this example all transations made on `MyPageExtension2` have translation ids as if they were on `MyPageExtension1`
```
pageextension 1 MyPageExtension1 extends "Customer Card" { }
pageextension 2 MyPageExtension2 extends "Customer Card" { }
```

Anyone got an idea on this?
